crates/accelerate/src/equivalence.rs

```diff
@@ -0,0 +1,786 @@
+        self.to_string()
+    }
+
+    fn __eq__(slf: &Bound<Self>, other: &Bound<PyAny>) -> PyResult<bool> {
+        Ok(slf.getattr("key")?.eq(other.getattr("key")?)?
+            && slf.getattr("equivs")?.eq(other.getattr("equivs")?)?)
+    }
+
+    fn __getnewargs__<'py>(
+        slf: &'py Bound<'py, Self>,
+    ) -> PyResult<(Bound<'py, PyAny>, Bound<'py, PyAny>)> {
+        Ok((slf.getattr("key")?, slf.getattr("equivs")?))
+    }
+}
+
+impl Display for NodeData {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "NodeData(key={}, equivs=[{}])",
+            self.key,
+            self.equivs.iter().format(", ")
+        )
+    }
+}
+
+#[pyclass(frozen, sequence, module = "qiskit._accelerate.equivalence")]
+#[derive(Debug, Clone)]
+pub struct EdgeData {
+    #[pyo3(get)]
+    pub index: usize,
+    #[pyo3(get)]
+    pub num_gates: usize,
+    #[pyo3(get)]
+    pub rule: Equivalence,
+    #[pyo3(get)]
+    pub source: Key,
+}
+
+#[pymethods]
+impl EdgeData {
+    #[new]
+    #[pyo3(signature = (index, num_gates, rule, source))]
+    fn new(index: usize, num_gates: usize, rule: Equivalence, source: Key) -> Self {
+        Self {
+            index,
+            num_gates,
+            rule,
+            source,
+        }
+    }
+
+    fn __repr__(&self) -> String {
+        self.to_string()
+    }
+
+    fn __eq__(slf: &Bound<Self>, other: &Bound<Self>) -> PyResult<bool> {
+        let other_borrowed = other.borrow();
+        let slf_borrowed = slf.borrow();
+        Ok(slf_borrowed.index == other_borrowed.index
+            && slf_borrowed.num_gates == other_borrowed.num_gates
+            && slf_borrowed.source == other_borrowed.source
+            && other.getattr("rule")?.eq(slf.getattr("rule")?)?)
+    }
+
+    fn __getnewargs__(slf: PyRef<Self>) -> (usize, usize, Equivalence, Key) {
+        (
+            slf.index,
+            slf.num_gates,
+            slf.rule.clone(),
+            slf.source.clone(),
+        )
+    }
+}
+
+impl Display for EdgeData {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "EdgeData(index={}, num_gates={}, rule={}, source={})",
+            self.index, self.num_gates, self.rule, self.source
+        )
+    }
+}
+
+/// Enum that helps extract the Operation and Parameters on a Gate.
+#[derive(Debug, Clone)]
+pub struct GateOper {
+    operation: PackedOperation,
+    params: SmallVec<[Param; 3]>,
+}
+
+impl<'py> FromPyObject<'py> for GateOper {
+    fn extract(ob: &'py PyAny) -> PyResult<Self> {
+        let op_struct: OperationFromPython = ob.extract()?;
+        Ok(Self {
+            operation: op_struct.operation,
+            params: op_struct.params,
+        })
+    }
+}
+
+/// Representation of QuantumCircuit which the original circuit object + an
+/// instance of `CircuitData`.
+#[derive(Debug, Clone)]
+pub struct CircuitRep(pub CircuitData);
```

Comment:
Why do you need to wrap `CircuitData` in `CircuitRep`?

Reply:
This is here just for seamless conversion to Python, since right now the `BasisTranslator` needs to receive circuits, although I will be ultimately changing this since my implementation of `BasisTranslator` is made to use `CircuitData` by default. I will add a `TODO` comment describing why this is necessary at the moment.